File: CodingInterviewsJavaSolutions/Remove Nth Node From End of List.java
class Solution {
    public int length(ListNode head) {
        int count = 0;
        while (head != null) {
            count++;
            head = head.next;
        }
        return count;
    }

    public ListNode getNth(ListNode head, int n) {
        int count = 0;
        ListNode cur = head;
        while (cur != null) {
            count++;
            if (count == n) {
                return cur;
            }
            cur = cur.next;
        }
        return null;
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        int len = length(head);
        head = new ListNode(-1, head);

        ListNode prev = getNth(head, len - n + 1);
        ListNode toRemove = prev.next;

        prev.next = prev.next.next;
        toRemove = null;

        toRemove = head;
        head = head.next;

        toRemove = null;
        return head;
    }
}

File: CodingInterviewsPythonSolutions/Remove Nth Node From End of List.py
class Solution:
    def length(self, head):
        count = 0
        while head:
            count += 1
            head = head.next
        return count

    def getNth(self, head, n):
        count = 0
        cur = head
        while cur:
            count += 1
            if count == n:
                return cur
            cur = cur.next
        return None

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        len = self.length(head)
        head = ListNode(-1, head)

        prev = self.getNth(head, len - n + 1)
        toRemove = prev.next

        prev.next = prev.next.next
        del toRemove

        toRemove = head
        head = head.next

        del toRemove
        return head
